baseServer.go

```diff
@@ -0,0 +1,216 @@
+	turns int
+	//lock which stops race conditions when updating agentStoppedTalkingMap 
+	agentMapRWMutex sync.RWMutex
+	//stops multiple sends to messagingFinished during a round
+	doneChannelOnce sync.Once
+}
+
+func (server *BaseServer[T]) HandleStartOfTurn(iter, round int) {
+	server.doneChannelOnce = sync.Once{}
+	server.messagingFinished = make(chan struct{}, 1)
+	fmt.Printf("Iteration %d, Round %d starting...\n", iter, round)
+
+}
+
+func (serv *BaseServer[T]) endAgentListeningSession() bool {
+
+	ctx, cancel := context.WithTimeout(context.Background(), serv.turnTimeout)
+	defer cancel()
+	select {
+	case <-serv.messagingFinished:
+		serv.agentStoppedTalkingMap = make(map[uuid.UUID]struct{})
+		return true
+
+	case <-ctx.Done():
+		serv.agentStoppedTalkingMap = make(map[uuid.UUID]struct{})
+		return false
+	}
+}
+
+func (server *BaseServer[T]) HandleEndOfTurn(iter, round int) {
+	server.endAgentListeningSession()
+	fmt.Printf("Iteration %d, Round %d finished.\n", iter, round)
+}
+
+func (server *BaseServer[T]) RunAgentLoop() {}
+
+func (server *BaseServer[T]) SendMessage(msg IMessage[T], receivers []uuid.UUID) {
+	for _, receiver := range receivers {
+		go msg.InvokeMessageHandler(server.agentMap[receiver])
+	}
+
+}
+
+func (serv *BaseServer[T]) AddAgent(agent T) {
+	serv.agentMap[agent.GetID()] = agent
+	serv.agentIdSet[agent.GetID()] = struct{}{}
+}
+
+func (serv *BaseServer[T]) ViewAgentIdSet() map[uuid.UUID]struct{} {
+	return serv.agentIdSet
+}
+
+func (serv *BaseServer[T]) AccessAgentByID(id uuid.UUID) T {
+	return serv.agentMap[id]
+}
+
+func (serv *BaseServer[T]) Initialise() {}
+
+func (serv *BaseServer[T]) Start() {
+	serv.checkHandler()
+	turns := serv.turns
+	iterations := serv.iterations
+	for i := 0; i < iterations; i++ {
+		for j := 0; j < turns; j++ {
+			serv.HandleStartOfTurn(i+1, j+1)
+			serv.roundRunner.RunTurn()
+			serv.HandleEndOfTurn(i+1, j+1)
+		}
+	}
+}
+
+func (serv *BaseServer[T]) GetAgentMap() map[uuid.UUID]T {
+	return serv.agentMap
+}
+
+func (serv *BaseServer[T]) agentStoppedTalking(id uuid.UUID) {
+	serv.agentMapRWMutex.Lock()
+	fmt.Println("sending stop talking request,id:", id)
+	serv.agentStoppedTalkingMap[id] = struct{}{}
+
+	if len(serv.agentStoppedTalkingMap) == len(serv.agentMap) {
+		serv.doneChannelOnce.Do(func() {
+			select {
+			case serv.messagingFinished <- struct{}{}:
+				close(serv.messagingFinished)
+			default:
+			}
+		})
+	}
+	serv.agentMapRWMutex.Unlock()
+}
+
+func (serv *BaseServer[T]) SetRunHandler(handler RoundRunner) {
+	serv.roundRunner = handler
+}
+
+func (serv *BaseServer[T]) checkHandler() {
+	if serv.roundRunner == nil {
+		panic("round running handler has not been set. Have you run SetRunHandler?")
+	}
+}
+
+func (serv *BaseServer[T]) RunTurn() {}
+
+func (bs *BaseServer[T]) RemoveAgent(agentToRemove T) {
+	delete(bs.agentMap, agentToRemove.GetID())
+}
+
+func (bs *BaseServer[T]) GetIterations() int {
+	return bs.iterations
+}
+
+func (bs *BaseServer[T]) RunRound() {}
```

Comment:
shouldn't this invoke the `RoundRunner.RunRound` method by default?

Reply:
If so, we can probably remove it, in place of just invoking the `RoundRunner`'s method

Reply:
how about we remove it entirely and add Start to the roundRunner interface allowing for a more customised gameloop if preferred

Reply:
not sure. it's arguably better to override the start method in the server, to be honest.

is start anything other than running the round loop for `I` iterations and turn loop for `j` iterations?